File: src/main/go/codility/challenges/TricoloredTowers.go
package challenges

import "fmt"

/*
 * Link: https://app.codility.com/programmers/task/tricolored_towers/
 * Level: Medium
 * Score: 100% correctness and performance
 */
func main() {
	fmt.Println(Solution([]string{"aab", "cab", "baa", "baa"}))        // 3
	fmt.Println(Solution([]string{"zzz", "zbz", "zbz", "dgf"}))        // 2
	fmt.Println(Solution([]string{"abc", "cba", "cab", "bac", "bca"})) // 3
}

func Solution(towers []string) int {
	possibilities := make(map[string]int)
	maxCount := 0

	for i := 0; i < len(towers); i++ {
		tower := towers[i]
		updatePossibilities(tower, tower, true, possibilities, &maxCount)
		updatePossibilities(tower, fmt.Sprintf("%c%c%c", tower[1], tower[0], tower[2]), false, possibilities, &maxCount)
		updatePossibilities(tower, fmt.Sprintf("%c%c%c", tower[0], tower[2], tower[1]), false, possibilities, &maxCount)
	}

	return maxCount
}

func updatePossibilities(tower, permutation string, countSelf bool, possibilities map[string]int, maxCount *int) {
	if tower == permutation && !countSelf {
		return
	}
	count := possibilities[permutation] + 1
	possibilities[permutation] = count
	*maxCount = max(count, *maxCount)
}

// god... does this language seriously force you to implement this yourself? Well... that's proof enough that not everything Google does
// is neat and nice to use!
func max(x, y int) int {
	if x < y {
		return y
	}
	return x
}

File: src/main/java/codility/challenges/TricoloredTowers.kt
package codility.challenges

import kotlin.math.max

/**
 * Link: https://app.codility.com/programmers/task/tricolored_towers/
 * Level: Medium
 * Score: 100% correctness and performance
 */
fun main() {
    println(solution(arrayOf("aab", "cab", "baa", "baa"))) // 3
    println(solution(arrayOf("zzz", "zbz", "zbz", "dgf"))) // 2
    println(solution(arrayOf("abc", "cba", "cab", "bac", "bca"))) // 3
}

fun solution(towers: Array<String>): Int {
    val possibilities = mutableMapOf<String, Int>()
    var maxCount = 0

    fun updatePossibilities(tower: String, permutation: String = tower, countSelf: Boolean = true) {
        if (tower == permutation && !countSelf) return
        val count = possibilities.merge(permutation, 1) { a, b -> a + b }!!
        maxCount = max(maxCount, count)
    }

    for (tower in towers) {
        updatePossibilities(tower)
        updatePossibilities(tower, "${tower[1]}${tower[0]}${tower[2]}", false)
        updatePossibilities(tower, "${tower[0]}${tower[2]}${tower[1]}", false)
    }

    return maxCount
}

